pkg/jwks/jwks.go

```diff
@@ -0,0 +1,261 @@
+}
+
+// getKeys reads the keys for a single HealthAuthority from its jwks server.
+func (mgr *JWKSManager) getKeys(ha *model.HealthAuthority) ([]byte, error) {
+	if len(ha.JwksURI) == 0 {
+		return nil, nil
+	}
+
+	req, err := http.NewRequestWithContext(mgr.ctxt, "GET", ha.JwksURI, nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating connection: %w", err)
+	}
+
+	var resp *http.Response
+	resp, err = http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("reading connection: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("resp (%v) != %v", resp.StatusCode, http.StatusOK)
+	}
+
+	var bytes []byte
+	bytes, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading: %w", err)
+	}
+	return bytes, nil
+}
+
+// parseKeys parses the json response, returning the pem encoded public keys,
+// and versions.
+func parseKeys(data []byte) ([]string, map[string]string, error) {
+	if len(data) == 0 {
+		return nil, nil, nil
+	}
+
+	var jwks []jwk.JWK
+	if err := json.Unmarshal(data, &jwks); err != nil {
+		return nil, nil, fmt.Errorf("unmarshal error: %w", err)
+	}
+
+	keys := make([]string, len(jwks))
+	versions := make(map[string]string, len(jwks))
+	for i := range jwks {
+		spec, err := jwks[i].ParseKeySpec()
+		if err != nil {
+			return nil, nil, fmt.Errorf("parse error: %w", err)
+		}
+
+		var encoded []byte
+		encoded, err = x509.MarshalPKIXPublicKey(spec.Key)
+		if err != nil {
+			return nil, nil, fmt.Errorf("marshal error: %w", err)
+		}
+		keys[i] = string(pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: encoded}))
+		versions[keys[i]] = spec.KeyID
+	}
+	return keys, versions, nil
+}
+
+// stripKey strips pem signature from a key.
+func stripKey(k string) string {
+	k = strings.ReplaceAll(k, "-----BEGIN PUBLIC KEY-----", "")
+	k = strings.ReplaceAll(k, "-----END PUBLIC KEY-----", "")
+	k = strings.ReplaceAll(k, "\n", "")
+	return k
+}
+
+// findKeyMods compares the read keys, and the keys already in the database,
+// and returns lists of the deadKey indices, new keys. (note the dead keys are
+// sorted.)
+//
+// NB: We don't depend on the keyID for this, we look at the actual pem strings,
+// and compare them.
+func findKeyMods(ha *model.HealthAuthority, rxKeys []string) (deadKeys []int, newKeys []string) {
+	// Figure out which keys are active, or deleted.
+	keySet := make(map[string]int, len(rxKeys))
+	for _, key := range rxKeys {
+		keySet[stripKey(key)] = -1
+	}
+	for i, key := range ha.Keys {
+		keyStr := stripKey(key.PublicKeyPEM)
+		if _, ok := keySet[keyStr]; !ok {
+			deadKeys = append(deadKeys, i)
+		} else {
+			keySet[keyStr] = i
+		}
+	}
+	for _, key := range rxKeys {
+		if keySet[stripKey(key)] == -1 {
+			newKeys = append(newKeys, key)
+		}
+	}
+	sort.Ints(deadKeys)
+	return
+}
+
+// updateHA updates HealthAuthority's keys.
+func (mgr *JWKSManager) updateHA(haDB *hadb.HealthAuthorityDB, ha *model.HealthAuthority, idx int) error {
```

Comment:
suggest using the lock infrastructure on this sever to ensure that this is only updated by one instance at a time and possibly to ensure they are spaced out appropriately

Reply:
obsolete.